Grid/qcd/utils/GaugeGroupTwoIndex.h

```diff
@@ -108,60 +178,53 @@ class GaugeGroupTwoIndex : public GaugeGroup<ncolour, group_name> {
     }
 
     if (Index < ncolour * (ncolour - 1) / 2) {
-      baseOffDiagonal(a[Index][0], a[Index][1], eij, group_name());
+      baseOffDiagonal(a[Index][0], a[Index][1], eij);
     } else {
       baseDiagonal(Index, eij);
     }
   }
+    
+  template <class cplx>
+   static void base(int Index, iGroupMatrix<cplx> &eij, GroupName::Sp) {
```

Comment:
This function is identical to `base(..., GroupName::Sp)` except for the one `if` statement. We should avoid the code duplication and move it back to the state before. If this is actually a performance hit, we should include an
```cpp
if (std::is_same<group_name, GroupName::Sp>::value) {...}
```
block around it. This *should* be an [`if constexpr`](https://en.cppreference.com/w/cpp/language/if) but that only came into the standard in '17. Nevertheless, I've checked and even at `-O0` the check and content of the block actually [never get instantiated](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGEs6SuADJ4DJgAcj4ARpjEIGYArKQADqgKhE4MHt6%2B/ilpGQIhYZEsMXGJtpj2jgJCBEzEBNk%2BflwBdpgOmfWNBMUR0bHxSQoNTS257bbj/aGDZcOJAJS2qF7EyOwc5gDMocjeWADUJrtuBACeyZgA%2BgTETIQKZ9gmGgCCewdHmKfnTjGxEwrFe7y%2BnwImBYyQMULOF2ujFYfwAKmDPkCvA5jh9TgB2KyfY4k44AN1QeHQxyoEGWBKJeNJzP4xGOEFCBGOeDOABENGcLNyEVwBbshdZrHh6SZCeDmQrSXgqOyxugQCA8ApbgoUQjUaRuYJXhqyWIvJgZXLiYrbSTleyeeddlbGXb3UrJbteVxBfKPQrZbz/R6gyH3WqNWgvFyEQjuf83PHzGYTAk3AwU36bYqwzmCcHPmHxeDwZzjiwngw6Qzw18Acbdthjkxbpzs0ySZGUOtY%2Bdk2YzK32%2BnM4OO8zh4IAHS05YdusI9DrKL0V4t27LryrzAT0nd6N9pPnU6D1tbndpjNZkv588r%2BizukL/G8jirWicBK8PwcLSkVBOCTSxrGOBR1k2P49h4UgCE0d9VgAaxABINH0ThJB/eCAM4XgFBANC4L/d9SDgWAkDQGE6FichKEo5JqLiYAuF2LhSCwMk8C2AA1PBMAAdwAeRuX8YJoWgoWIfCICibColCRpLk4GD5OYYhLkEqJtC6IiYMothBEEhhaCU4j2MwKIvGANwxFofDuF4LBKyMcQzPwYFujJTB7P/TBVC6GNthgzlqmw2g8CiR51I8LBsIePAWGUkiqAMYAFF4gThMYJKZEEEQxHYKRcvkJQ1Gw3Q2IMIwUElSx9Ai/DIFWVBklqBh7IAWg6tU%2BVMEDLCkY4OsE/9UC84hiCpbz4FWTpumcCBXEmNpAgYdABlKco9FSdI2uW7aCjajahjiNi5ra3oJk8Vo9HOnpZmOxZTpmPp9rOh75k24YuFmiCtgkD8vywszAI4Y5VAADgANg6qHJGOYBkGQY4WOnLh2VwQgSFPVjll4IitGWJCULQz8OEw0hf1G3DbAI2D4KJ9CODMYHqY4fGGdWcb0mcSQgA), so no performance hit unless needed.

Reply:
You mean it's identical to `base(..., GroupName::SU)`?

Why can go back indeed. It's good the performance is not affected. Notice there is also a difference in the if statement splitting between diagonal and off diagonal. These two differences, coupled to a possible future extension to SO, and avoiding c++ > 14, made me opt for the proposed implementation rather the the previous one.

Reply:
Okay. Yes, I didn't spot the latter one. Is the first part about filling `a` agnostic of the group? In that case, the correct abstraction would rather be a function `numOffDiagonalBaseElements(group_name())` (or `static const` or the like) that makes this distinction explicit.